pkg/v1/v1util/matcher.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1util
+
+import (
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// ManifestMatcher function that is given a v1.Descriptor, and returns whether or
+// not it matches a given rule. Can match on anything it wants in the Descriptor.
+type ManifestMatcher func(desc v1.Descriptor) bool
+
+// NameMatcher returns a ManifestMatcher that matches on the name tag in
+// github.com/opencontainers/image-spec/specs-go/v1.AnnotationRefName
+func NameMatcher(name string) ManifestMatcher {
+	return AnnotationMatcher(imagespec.AnnotationRefName, name)
+}
+
+// AnnotationMatcher returns a ManifestMatcher that matches based on the provided annotation
+func AnnotationMatcher(key, value string) ManifestMatcher {
+	return func(desc v1.Descriptor) bool {
+		if desc.Annotations == nil {
+			return false
+		}
+		if aValue, ok := desc.Annotations[key]; ok && aValue == value {
+			return true
+		}
+		return false
+	}
+}
+
+// PlatformMatcher returns a ManifestMatcher that matches on the provided platform.
+// Ignores any descriptors that do not have a platform.
+func PlatformMatcher(platform v1.Platform) ManifestMatcher {
+	return func(desc v1.Descriptor) bool {
+		if desc.Platform == nil {
+			return false
+		}
+		return platformEqual(*desc.Platform, platform)
+	}
+}
+
+// MediaTypeMatcher returns a ManifestMatcher that matches based on one of the provided media types
+func MediaTypeMatcher(mediaTypes []string) ManifestMatcher {
```

Comment:
If we made this variadic, it would be nicer when trying to match a single media type, but annoying when trying to match multiple. Given that with the OCI spec there are basically two of each media type, it makes some sense to me to keep this as an array, but I want to point it out.

Reply:
Sure.